src/connectors/impls/gbq/writer/sink.rs

```diff
@@ -0,0 +1,660 @@
+            tag,
+            &val.as_i64()
+                .ok_or_else(|| ErrorKind::BigQueryTypeMismatch("i64", val.value_type()))?,
+            result,
+        ),
+        TableType::Bool => prost::encoding::bool::encode(
+            tag,
+            &val.as_bool()
+                .ok_or_else(|| ErrorKind::BigQueryTypeMismatch("bool", val.value_type()))?,
+            result,
+        ),
+        TableType::String
+        | TableType::Date
+        | TableType::Time
+        | TableType::Datetime
+        | TableType::Timestamp
+        | TableType::Numeric
+        | TableType::Bignumeric
+        | TableType::Geography => {
+            prost::encoding::string::encode(
+                tag,
+                &val.as_str()
+                    .ok_or_else(|| ErrorKind::BigQueryTypeMismatch("string", val.value_type()))?
+                    .to_string(),
+                result,
+            );
+        }
+        TableType::Struct => {
+            let mut struct_buf: Vec<u8> = vec![];
+            for (k, v) in val
+                .as_object()
+                .ok_or_else(|| ErrorKind::BigQueryTypeMismatch("object", val.value_type()))?
+            {
+                let subfield_description = field.subfields.get(&k.to_string());
+
+                if let Some(subfield_description) = subfield_description {
+                    encode_field(v, subfield_description, &mut struct_buf)?;
+                } else {
+                    warn!(
+                        "Passed field {} as struct field, not present in definition",
+                        k
+                    );
+                }
+            }
+            prost::encoding::encode_key(tag, WireType::LengthDelimited, result);
+            prost::encoding::encode_varint(struct_buf.len() as u64, result);
+            result.append(&mut struct_buf);
+        }
+        TableType::Bytes => {
+            prost::encoding::bytes::encode(
+                tag,
+                &Vec::from(
+                    val.as_bytes().ok_or_else(|| {
+                        ErrorKind::BigQueryTypeMismatch("bytes", val.value_type())
+                    })?,
+                ),
+                result,
+            );
+        }
+        TableType::Json => {
+            warn!("Found a field of type JSON, this is not supported, ignoring.");
+        }
+        TableType::Interval => {
+            warn!("Found a field of type Interval, this is not supported, ignoring.");
+        }
+
+        TableType::Unspecified => {
+            warn!("Found a field of unspecified type - ignoring.");
+        }
+    }
+
+    Ok(())
+}
+
+impl JsonToProtobufMapping {
+    pub fn new(vec: &Vec<TableFieldSchema>) -> Self {
+        let descriptor = map_field("table", vec);
+
+        Self {
+            descriptor: descriptor.0,
+            fields: descriptor.1,
+        }
+    }
+
+    pub fn map(&self, value: &Value) -> Result<Vec<u8>> {
+        let mut result = vec![];
```

Comment:
it is usually more effective to pre-allocate a chunk, even if it is too big, to avoid reallocations while filling the buffer.
One thing we could do is to store the final capacity of the created vec and use this for the next one too, assuming the value sizes will remain roughly the same.